x-pack/plugins/endpoint/server/services/resolver/search_handler.ts

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { IScopedClusterClient } from 'kibana/server';
+import { CountResponse } from 'elasticsearch';
+import {
+  ResolverLegacyData,
+  ResolverElasticEndpointData,
+  ResolverData,
+  BaseResult,
+  ResolverResponse,
+} from '../../../common/types';
+import { EndpointAppContext, JSONish, Total } from '../../types';
+import { ResolverNode, ResolverDataHit } from './common';
+import { ResolverLegacyNode } from './legacy_node';
+import { ResolverElasticEndpointNode } from './es_endpoint_node';
+import { getPagination, PaginationInfo } from './query_builder';
+
+export interface ParentAndResolverData {
+  parentEntityID: string;
+  events: ResolverData[];
+}
+
+export abstract class BaseSearchHandler {
+  constructor(
+    protected readonly client: IScopedClusterClient,
+    protected readonly endpointContext: EndpointAppContext,
+    protected readonly pageInfo: PaginationInfo,
+    protected readonly countQuery: JSONish,
+    protected readonly entityID: string
+  ) {}
+
+  protected static isLegacyData(
+    data: ResolverLegacyData | ResolverElasticEndpointData
+  ): data is ResolverLegacyData {
+    return (data as ResolverLegacyData).endgame?.unique_pid !== undefined;
+  }
+
+  protected static nodeCreator(data: ResolverData): ResolverNode {
+    if (BaseSearchHandler.isLegacyData(data)) {
+      return new ResolverLegacyNode(data);
+    }
+    return new ResolverElasticEndpointNode(data);
+  }
+  protected async buildPagination(total: Total): Promise<BaseResult> {
+    const { page, pageSize, from } = await getPagination(this.endpointContext, this.pageInfo);
+    let amount = total.value;
+    if (total.relation === 'gte') {
```

Comment:
Since this means we will always be looking for the exact total number of hits, why not use the track_total_hits option in the original request for the documents and avoid having the logic for conditionally counting the hits using a separate search?
https://www.elastic.co/guide/en/elasticsearch/reference/7.5/search-request-body.html#request-body-search-track-total-hits

I think this would let us remove a lot of the code and make it much simpler.